### PR TITLE
Resolving various MSVC warnings

### PR DESCRIPTION
**Pull request by olgavrou**
This addresses MSVC warnings for `fopen`, `strtok`, `strcpy`, `fscanf`, `fprintf`, `getenv` being unsafe, and potentially uninitialized variable warnings.

The only warnings left are `C4127: conditional expression is constant` and `C4996: 'gethostbyname': Use getaddrinfo() instead`

### CODE REVIEW

File: library/recommend.cc [hunk 1]
```diff
@@ -141,22 +141,28 @@ int main(int argc, char* argv[])
...
   FILE* fU;
+  int errnofu;
   FILE* fI;
+  int errnofi;
 
-  if ((fB = fopen(blacklistfilename.c_str(), "r")) == NULL)
+  VW::file_open(&fB, blacklistfilename.c_str(), "r", &errnofb);
+  if (errnofb != 0)
   {
     fprintf(stderr, "can't open %s: %s\n", blacklistfilename.c_str(), strerror(errno));
```
Comment (lalo) [inline library/recommend.cc:153]: @jackgerrits should this be using VW::strerror_to_string?

Comment (lalo) [reply library/recommend.cc:153]: not relevant to this pr though

Comment (olgavrou) [reply library/recommend.cc]: Replaced.

File: vowpalwabbit/memory_tree.cc [hunk 2]
```diff
@@ -376,6 +377,8 @@ inline int random_sample_example_pop(memory_tree& b, uint64_t& cn)
 
 // train the node with id cn, using the statistics stored in the node to
 // formulate a binary classificaiton example.
+VW_WARNING_STATE_PUSH
+VW_WARNING_DISABLE_POTENTIALLY_UNINITIALIZED
```
[the remark this answers is not in the document]
Comment (olgavrou) [reply vowpalwabbit/memory_tree.cc:381]:
```
MULTICLASS::label_t mc;
MULTILABEL::labels multilabels;
MULTILABEL::labels preds;
```

the last two are structs with a `v_array` member and are used in if/else loops which causes the compiler to complain.

[the remark this answers is not in the document]
Comment (jackgerrits) [reply vowpalwabbit/memory_tree.cc:381]: Seems fishy to me. They are relying on being initialized with zero memory for it to function. We can come back to this though, the good thing about `VW_WARNING_DISABLE_POTENTIALLY_UNINITIALIZED` being in the code is we can audit it.

[the remark this answers is not in the document]
Comment (olgavrou) [reply vowpalwabbit/memory_tree.cc:381]: discussed offline, will call `v_init` instead and remove `VW_WARNING_DISABLE_POTENTIALLY_UNINITIALIZED `

File: vowpalwabbit/crossplat_compat.h [hunk 3]
```diff
@@ -4,14 +4,50 @@
...
+  // strcpy_s returns an errno_t
+  return strcpy_s(dest, dest_size, src);
+#else
+  (void)dest_size; // unused here
+  strcpy(dest, src);
+  return 0;
+#endif
+}
+
+inline int file_open(FILE **pf, const char *filename, const char *mode)
```
Comment (jackgerrits) [inline vowpalwabbit/crossplat_compat.h:39]: Should this new function be used in the file io adapters?
https://github.com/VowpalWabbit/vowpal_wabbit/blob/d4476ff327dc9416915b456b78d17da032c66a44/vowpalwabbit/io/io_adapter.cc#L241

Comment (olgavrou) [reply vowpalwabbit/crossplat_compat.h:39]: It looks like the io_adapter uses `_sopen_s/open` to get a file descriptor and do a lower level read/write of the underlying file vs `fopen_s/fopen` which provide for example buffering, so I am not sure they are suitable here?

Comment (jackgerrits) [reply vowpalwabbit/crossplat_compat.h:39]: Got it, no prob

**Pull request merged**